tests/backends/data/test_clickhouse.py

```diff
@@ -0,0 +1,643 @@
+    )
+    backend = ClickHouseDataBackend(settings)
+    backend._bulk_import(documents)  # pylint: disable=protected-access
+
+    # Test filtering
+    query = ClickHouseQuery(where_clause="event.bool = 1")
+    results = list(backend.read(query=query))
+    assert len(results) == 2
+    assert results[0]["event"] == statements[0]
+    assert results[1]["event"] == statements[2]
+
+    # Test fields
+    query = ClickHouseQuery(return_fields=["event_id", "event.bool"])
+    results = list(backend.read(query=query))
+    assert len(results) == 3
+    assert len(results[0]) == 2
+    assert results[0]["event_id"] == documents[0][0]
+    assert results[0]["event.bool"] == statements[0]["bool"]
+    assert results[1]["event_id"] == documents[1][0]
+    assert results[1]["event.bool"] == statements[1]["bool"]
+    assert results[2]["event_id"] == documents[2][0]
+    assert results[2]["event.bool"] == statements[2]["bool"]
+
+    # Test filtering and projection
+    query = ClickHouseQuery(
+        where_clause="event.bool = 0", return_fields=["event_id", "event.bool"]
+    )
+    results = list(backend.read(query=query))
+    assert len(results) == 1
+    assert len(results[0]) == 2
+    assert results[0]["event_id"] == documents[1][0]
+    assert results[0]["event.bool"] == statements[1]["bool"]
+
+
+def test_backends_data_clickhouse_data_backend_data_backend_list_method(clickhouse):
+    """Test the clickhouse backend list method."""
+
+    settings = ClickHouseDataBackend.settings_class(
+        HOST=CLICKHOUSE_TEST_HOST,
+        PORT=CLICKHOUSE_TEST_PORT,
+        DATABASE=CLICKHOUSE_TEST_DATABASE,
+        EVENT_TABLE_NAME=CLICKHOUSE_TEST_TABLE_NAME,
+    )
+    backend = ClickHouseDataBackend(settings)
+
+    assert list(backend.list(details=True)) == [{"name": CLICKHOUSE_TEST_TABLE_NAME}]
+    assert list(backend.list(details=False)) == [CLICKHOUSE_TEST_TABLE_NAME]
+
+
+# pylint: disable=protected-access
+def test_backends_data_clickhouse_data_backend__to_documents_method_invalid_timestamp(
+    clickhouse,
+):
+    """Test the clickhouse backend _to_documents method with an invalid timestamp."""
+    valid_timestamp = (datetime.now() - timedelta(seconds=3)).isoformat()
+    invalid_timestamp = "This is not a valid timestamp!"
+    invalid_statement = {
+        "id": str(uuid.uuid4()),
+        "bool": 0,
+        "timestamp": invalid_timestamp,
+    }
+
+    statements = [
+        {"id": str(uuid.uuid4()), "bool": 1, "timestamp": valid_timestamp},
+        invalid_statement,
+    ]
+    documents = ClickHouseDataBackend._to_documents(statements, ignore_errors=False)
+
+    assert str(next(documents)[0]) == statements[0]["id"]
+
+    with pytest.raises(
+        BadFormatException,
+        match=f"Statement has an invalid or missing id or "
+        f"timestamp field: {invalid_statement}",
+    ):
+        next(documents)
+
+
+# pylint: disable=protected-access
+def test_backends_data_clickhouse_data_backend__to_documents_method():
```

Comment:
Is a double underscore usual when testing private methods? 
Otherwise, I would propose remove it
```suggestion
def test_backends_data_clickhouse_data_backend_to_documents_method():
```